File: backend/app/core/risk_engine.py
"""Scoring — the heart of the system (F5, F7, F8).

Pure functions only: no I/O, no network, no clock, no globals beyond the rules
singleton. Data in, data out. That is what makes the six test vectors trivial to
assert and makes a rule change safe.

The engine never sees the APK. It receives an already-extracted group set, so
its cost is independent of file size and the same input always yields the same
score.
"""

from __future__ import annotations

from dataclasses import dataclass

from app.core import patterns as pattern_module
from app.core.category import mismatch_penalty_disabled
from app.core.permissions import permission_items, present_groups
from app.core.rules_loader import Band, Rules, get_rules
from app.models.schemas import (
    BreakdownItem,
    CategoryResult,
    ParsedApk,
    PatternMatch,
    PermissionItem,
)


@dataclass(frozen=True)
class ScoreResult:
    score: int
    band: Band
    verdict: str
    recommendation: str
    critical_override: bool
    breakdown: tuple[BreakdownItem, ...]
    permissions: tuple[PermissionItem, ...]
    patterns: tuple[PatternMatch, ...]
    present_groups: frozenset[str]
    unexpected_groups: frozenset[str]


def band_for(score: int, rules: Rules) -> Band:
    for band in rules.bands:
        if band.contains(score):
            return band
    # Only reachable if the bands in rules.json have a gap; clamp to the top.
    return rules.bands[-1]


def score_apk(
    apk: ParsedApk,
    category: CategoryResult,
    rules: Rules | None = None,
    impersonation_detected: bool = False,
    repackaging_detected: bool = False,
) -> ScoreResult:
    """Score an APK against its declared category.

    Order of operations follows the spec exactly: group points first, then the
    two penalties, then pattern bonuses, then the cap and band, with a critical
    override applied last.
    """
    rules = rules or get_rules()

    expected = rules.categories[category.id].expected_groups
    groups = present_groups(apk.permissions, apk.services, apk.receivers, rules)

    breakdown: list[BreakdownItem] = []
    score = 0

    for group in sorted(groups):
        points = rules.group_points.get(group, 0)
        is_expected = group in expected

        if points == 0 and not is_expected:
            # Zero-point groups (boot) only exist to feed patterns. They are
            # still reported so the reader can see the app declares them.
            breakdown.append(
                BreakdownItem(
                    rule=f"group:{group}",
                    points=0,
                    reason_key="perm.info",
                    status="info",
                    params={"group": group},
                )
            )
            continue

        awarded = 0 if is_expected else points
        score += awarded

        breakdown.append(
            BreakdownItem(
                rule=f"group:{group}",
                points=awarded,
                reason_key="perm.expected" if is_expected else "perm.unexpected",
                status="expected" if is_expected else "unexpected",
                params={"group": group, "category": category.id},
            )
        )

    unexpected = frozenset(group for group in groups if group not in expected)

    if len(unexpected) > rules.many_sensitive_threshold:
        score += rules.many_sensitive_points
        breakdown.append(
            BreakdownItem(
                rule="many_sensitive",
                points=rules.many_sensitive_points,
                reason_key="rule.many_sensitive",
                status="unexpected",
                params={"count": str(len(unexpected))},
            )
        )

    if unexpected and not mismatch_penalty_disabled(category.id, rules):
        score += rules.mismatch_points
        breakdown.append(
            BreakdownItem(
                rule="mismatch",
                points=rules.mismatch_points,
                reason_key="rule.mismatch",
                status="unexpected",
                params={"category": category.id},
            )
        )

    matched = pattern_module.matched_patterns(
        groups, category.id, apk.has_launcher_activity, rules
    )
    for match in matched:
        score += match.bonus
        breakdown.append(
            BreakdownItem(
                rule=f"pattern:{match.id}",
                points=match.bonus,
                reason_key=match.reason_key,
                status="critical" if match.critical else "unexpected",
                params={},
            )
        )

    if impersonation_detected:
        score += rules.impersonation_points
        breakdown.append(
            BreakdownItem(
                rule="impersonation",
                points=rules.impersonation_points,
                reason_key="impersonation.detected",
                status="critical",
                params={},
            )
        )

    score = min(score, rules.max_score)
    band = band_for(score, rules)

    critical = (
        pattern_module.has_critical_pattern(matched)
        or impersonation_detected
        or repackaging_detected
    )

    verdict = band.verdict
    recommendation = band.recommendation
    if critical and verdict != "red":
        verdict = "red"
        recommendation = "do_not_install"

    return ScoreResult(
        score=score,
        band=band,
        verdict=verdict,
        recommendation=recommendation,
        critical_override=critical,
        breakdown=tuple(breakdown),
        permissions=tuple(permission_items(apk.permissions, expected, rules)),
        patterns=tuple(matched),
        present_groups=groups,
        unexpected_groups=unexpected,
    )

File: shared/src/commonMain/kotlin/com/appautopsy/analysis/core/RiskEngine.kt
package com.appautopsy.analysis.core

import com.appautopsy.analysis.model.Band
import com.appautopsy.analysis.model.BreakdownItem
import com.appautopsy.analysis.model.CategoryResult
import com.appautopsy.analysis.model.ParsedApk
import com.appautopsy.analysis.model.ScoreResult
import com.appautopsy.analysis.model.Verdict
import com.appautopsy.analysis.rules.Rules

/**
 * Scoring: the heart of the system.
 *
 * Pure functions only. No I/O, no network, no clock, no globals. Data in, data
 * out. That is what makes the test vectors trivial to assert and makes a rule
 * change safe.
 *
 * The engine never sees the APK. It receives an already-extracted group set, so
 * its cost is independent of file size and the same input always yields the
 * same score.
 */
object RiskEngine {

    /**
     * Score an APK against its declared category.
     *
     * Order of operations is fixed: group points first, then the two penalties,
     * then pattern bonuses, then impersonation, then the cap and band, with the
     * critical override applied last.
     *
     * Time O(G + K), space O(G), where G is the group set (about 15 max) and K
     * the pattern count. Constant with respect to APK size.
     */
    fun score(
        apk: ParsedApk,
        category: CategoryResult,
        rules: Rules,
        impersonationDetected: Boolean = false,
        repackagingDetected: Boolean = false,
    ): ScoreResult {
        val expected = Category.expectedGroups(category.id, rules)
        val groups = Permissions.presentGroups(
            apk.permissions, apk.services, apk.receivers, rules,
        )

        val breakdown = mutableListOf<BreakdownItem>()
        var score = 0

        for (group in groups.sorted()) {
            val points = rules.groupPoints[group] ?: 0
            val isExpected = group in expected

            if (points == 0 && !isExpected) {
                // Zero-point groups (boot) only exist to feed patterns. They are
                // still reported so the reader can see the app declares them.
                breakdown += BreakdownItem(
                    rule = "group:$group",
                    points = 0,
                    reasonKey = "perm.info",
                    status = "info",
                    params = mapOf("group" to group),
                )
                continue
            }

            val awarded = if (isExpected) 0 else points
            score += awarded

            breakdown += BreakdownItem(
                rule = "group:$group",
                points = awarded,
                reasonKey = if (isExpected) "perm.expected" else "perm.unexpected",
                status = if (isExpected) "expected" else "unexpected",
                params = mapOf("group" to group, "category" to category.id),
            )
        }

        val unexpected = groups.filter { it !in expected }.toSet()

        if (unexpected.size > rules.manySensitiveThreshold) {
            score += rules.manySensitivePoints
            breakdown += BreakdownItem(
                rule = "many_sensitive",
                points = rules.manySensitivePoints,
                reasonKey = "rule.many_sensitive",
                status = "unexpected",
                params = mapOf("count" to unexpected.size.toString()),
            )
        }

        // The mismatch penalty fires once, however many unexpected groups exist.
        if (unexpected.isNotEmpty() && !Category.mismatchPenaltyDisabled(category.id, rules)) {
            score += rules.mismatchPoints
            breakdown += BreakdownItem(
                rule = "mismatch",
                points = rules.mismatchPoints,
                reasonKey = "rule.mismatch",
                status = "unexpected",
                params = mapOf("category" to category.id),
            )
        }

        val matched = Patterns.matched(groups, category.id, apk.hasLauncherActivity, rules)
        for (match in matched) {
            score += match.bonus
            breakdown += BreakdownItem(
                rule = "pattern:${match.id}",
                points = match.bonus,
                reasonKey = match.reasonKey,
                status = if (match.critical) "critical" else "unexpected",
                params = emptyMap(),
            )
        }

        if (impersonationDetected) {
            score += rules.impersonationPoints
            breakdown += BreakdownItem(
                rule = "impersonation",
                points = rules.impersonationPoints,
                reasonKey = "impersonation.detected",
                status = "critical",
                params = emptyMap(),
            )
        }

        score = minOf(score, rules.maxScore)
        val band = bandFor(score, rules)

        val critical = Patterns.hasCritical(matched) ||
            impersonationDetected ||
            repackagingDetected

        // A critical finding forces red regardless of the numeric score. An app
        // that acts like a banking trojan is not "mostly fine" at 60 points.
        var verdict = Verdict.fromId(band.verdict)
        var recommendation = band.recommendation
        if (critical && verdict != Verdict.RED) {
            verdict = Verdict.RED
            recommendation = "do_not_install"
        }

        return ScoreResult(
            score = score,
            band = Band.fromId(band.id),
            verdict = verdict,
            recommendation = recommendation,
            criticalOverride = critical,
            breakdown = breakdown,
            permissions = Permissions.permissionItems(apk.permissions, expected, rules),
            patterns = matched,
            presentGroups = groups,
            unexpectedGroups = unexpected,
        )
    }

    /**
     * The band a score falls into.
     *
     * Time O(B) over a fixed three bands. A gap in the ranges falls back to the
     * top band, which fails safe: an impossible score is treated as dangerous
     * rather than as safe.
     */
    fun bandFor(score: Int, rules: Rules): com.appautopsy.analysis.rules.BandDef =
        rules.bands.firstOrNull { it.contains(score) } ?: rules.bands.last()
}